### PR TITLE
Allow AddTask to be called from an external thread

### PR DESCRIPTION
**Pull request by cwfitzgerald**
This is a fairly simple change that allows thread that are not part of the thread pool to still add tasks. Waiting on counters is another issue that is going to be much harder to solve, but for now, this works.

### CODE REVIEW

File: source/task_scheduler.cpp [hunk 1]
```diff
@@ -332,12 +332,21 @@ void TaskScheduler::Run(uint fiberPoolSize, TaskFunction mainTask, void *mainTas
...
 	}
 
 	const TaskBundle bundle = {task, counter};
-	m_tls[GetCurrentThreadIndex()].TaskQueue.Push(bundle);
+	std::size_t const threadIdx = GetCurrentThreadIndex();
+	if (threadIdx != FTL_INVALID_INDEX) {
+		m_tls[GetCurrentThreadIndex()].TaskQueue.Push(bundle);
+	}
+	else {
+		m_tls[externalIndex].TaskQueue.Push(bundle);
```
Comment (RichieSams) [inline source/task_scheduler.cpp]: TaskQueue.Push isn't thread safe. The internal algorithm has strong assumptions that there is a single pusher, and multiple stealers.

**Comment by cwfitzgerald** [comment]
Ah gotcha. This presents an interesting problem. How do we easily allow AddTask and friends to be called from outside a fiber? The reason this is an issue is if you are middleware, you get called from some other thread and need to add tasks to the work pool from that outside thread.

**Comment by martty** [comment]
I suppose you could add another queue (representing the external thread), and you can add work to that, which the other threads would steal and do. Of course this requires sync on the external queue (as mentioned, WFQ's Push cannot be called concurrently).

**Comment by cwfitzgerald** [comment]
Pinging @dethraid as I made this change for him.

**Comment by DethRaid** [comment]
Just got home from work. Here's what's up:

I'm making a middleware renderer that's going to be embedded into existing games. For now I'm just putting it into Minecraft, but I'd like to see how many other games I can shove it into - Fallout: New Vegas, Battlefront 2 2005, Skyrim, etc

I don't have any control over or even knowledge of the existing game's `main` function. The intended usage pattern of Fiber Tasking Library seems to be that `ftl::TaskScheduler::Run` executes your `main` function - `Run` creates the thread pool and fiber pool, executes all the work you give it, and cleans up the thread pool and fiber pool before returning. This usage pattern is not possible for me. I can't (well, I really don't want to) wrap the existing games' `main` functions

The solution @cwfitzgerald and I came up with was to call `ftl::TaskScheduler::Run` in a separate thread. That thread can spin or sleep or something until my program exits. In this usage pattern. `Run` isn't called from the main thread, but I need to give work to my task scheduler instance from the main thread. There's a few ways that could be accomplished, and the code in this PR seemed like the simplest

That is the use case of this feature

**Pull request closed without merging**